src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -40,44 +39,52 @@ internal partial class TextDocumentState
 
         protected TextDocumentState(
             SolutionServices solutionServices,
+            IDocumentServiceProvider documentServiceProvider,
             DocumentInfo.DocumentAttributes attributes,
             SourceText sourceTextOpt,
             ValueSource<TextAndVersion> textAndVersionSource,
             ValueSource<DocumentStateChecksums> lazyChecksums)
         {
             this.solutionServices = solutionServices;
-            this.attributes = attributes;
             this.sourceTextOpt = sourceTextOpt;
             this.textAndVersionSource = textAndVersionSource;
 
+            Attributes = attributes;
+            DocumentServices = documentServiceProvider;
+
             // for now, let it re-calculate if anything changed.
             // TODO: optimize this so that we only re-calcuate checksums that are actually changed
             _lazyChecksums = new AsyncLazy<DocumentStateChecksums>(ComputeChecksumsAsync, cacheResult: true);
         }
 
+        public DocumentInfo.DocumentAttributes Attributes { get; }
+
+        /// <summary>
+        /// A <see cref="IDocumentServiceProvider"/> associated with this document
+        /// 
+        /// not every <see cref="Workspace"/> or <see cref="Document"/> supports 
```

Comment:
Should we just sub in a no-op provider that always returns null?

Reply:
yep. I agree.

Reply:
ah. yes.  I will change the comment.